mlir/unittests/Analysis/Presburger/ParserTest.cpp

```diff
@@ -0,0 +1,160 @@
+  str = "(i)[] : () )";
+
+  set = parsePresburgerSet(str, &ctx);
+  EXPECT_TRUE(failed(set));
+
+  // `and` is only allowed inside a convex set.
+  str = "(i)[] : (i <= 2) and (i >= 3)";
+
+  set = parsePresburgerSet(str, &ctx);
+  EXPECT_TRUE(failed(set));
+
+  // Reused variable names are not supported
+  str = "(i,i)[] : (i <= 2) or (i >= 3)";
+
+  set = parsePresburgerSet(str, &ctx);
+  EXPECT_TRUE(failed(set));
+
+  str = "(i)[i] : (i <= 2) or (i >= 3)";
+
+  set = parsePresburgerSet(str, &ctx);
+  EXPECT_TRUE(failed(set));
+}
+
+TEST(ParserTest, simpleEq) {
+  MLIRContext ctx;
+  auto str = "(i) : (i = 0)";
+
+  FailureOr<PresburgerSet> set = parsePresburgerSet(str, &ctx);
+  EXPECT_TRUE(succeeded(set));
+
+  PresburgerSet ex =
+      makeSetFromFACs(1, 0, {makeFACFromConstraints(1, 0, {}, {{1, 0}})});
+  EXPECT_TRUE(set->isEqual(ex));
+}
+
+TEST(ParserTest, simpleIneq) {
+  MLIRContext ctx;
+  auto str = "(i)[] : (i >= 0)";
+
+  FailureOr<PresburgerSet> set = parsePresburgerSet(str, &ctx);
+  EXPECT_TRUE(succeeded(set));
+
+  PresburgerSet ex = makeSetFromFACs(1, 0, {makeFACFromIneqs(1, 0, {{1, 0}})});
+  EXPECT_TRUE(set->isEqual(ex));
+}
+
+TEST(ParserTest, simpleAnd) {
+  MLIRContext ctx;
+  auto str = "(i)[] : (i >= 0 and i <= 3)";
+
+  FailureOr<PresburgerSet> set = parsePresburgerSet(str, &ctx);
+  EXPECT_TRUE(succeeded(set));
+
+  PresburgerSet ex =
+      makeSetFromFACs(1, 0, {makeFACFromIneqs(1, 0, {{1, 0}, {-1, 3}})});
+  EXPECT_TRUE(set->isEqual(ex));
+}
+
+TEST(ParserTest, simpleOr) {
+  MLIRContext ctx;
+  auto str = "(i)[] : (i >= 0) or (i <= 3)";
+
+  FailureOr<PresburgerSet> set = parsePresburgerSet(str, &ctx);
+  EXPECT_TRUE(succeeded(set));
+
+  PresburgerSet ex = makeSetFromFACs(
+      1, 0,
+      {makeFACFromIneqs(1, 0, {{1, 0}}), makeFACFromIneqs(1, 0, {{-1, 3}})});
+  EXPECT_TRUE(set->isEqual(ex));
+}
+
+TEST(ParserTest, higherDim) {
+  MLIRContext ctx;
+  auto str = "(x,y)[] : (x >= 2 and y >= 2 and x <= 10 and y <= 10 and x + y "
+             ">= 2 and x + y <= 30 and x - y >= 0 and x -y <= 10)";
+
+  FailureOr<PresburgerSet> set = parsePresburgerSet(str, &ctx);
+  EXPECT_TRUE(succeeded(set));
+
+  PresburgerSet ex =
+      makeSetFromFACs(2, 0,
+                      {makeFACFromIneqs(2, 0,
+                                        {
+                                            {1, 0, -2},   // x >= 2.
+                                            {0, 1, -2},   // y >= 2.
+                                            {-1, 0, 10},  // x <= 10.
+                                            {0, -1, 10},  // y <= 10.
+                                            {1, 1, -2},   // x + y >= 2.
+                                            {-1, -1, 30}, // x + y <= 30.
+                                            {1, -1, 0},   // x - y >= 0.
+                                            {-1, 1, 10},  // x - y <= 10.
+                                        })});
+  EXPECT_TRUE(set->isEqual(ex));
+}
+
```

Comment:
Some more possible test cases

`(x) : ` (invalid)
`(x) : (x >= 0 and x <= 5) or (x >= -2 and x <= -5)`
`(x) : (x - >= -2)` (invalid)
`(x) : (1x >= -)` (invalid)
`(x) : (-x - -2 >= - -2)` (invalid)
`(x) : (-1x + -2 >= +2)`
`(x) : (1x -2 >= -2)`

Reply:
Currently, the set `(x) : (-1x + -2 >= +2)` cannot be parsed, as Parser::parseOptionalInteger does not allow a leading `+`.

Reply:
I see, okay.